core/templates/dev/head/exploration_editor/ExplorationEditor.js

```diff
@@ -620,49 +624,140 @@ oppia.controller('ExplorationSaveAndPublishButtons', [
+          !explorationCategoryService.savedMemento ||
+          !explorationLanguageCodeService.savedMemento) {
+         $modal.open({
+           templateUrl: 'modals/addExplorationMetadata',
+           backdrop: true,
+           controller: [
+             '$scope', '$modalInstance', 'explorationObjectiveService',
+             'explorationTitleService', 'explorationCategoryService',
+             'explorationStatesService', 'CATEGORY_LIST',
+             'explorationLanguageCodeService',
+             function($scope, $modalInstance, explorationObjectiveService,
+             explorationTitleService, explorationCategoryService,
+             explorationStatesService, CATEGORY_LIST,
+             explorationLanguageCodeService) {
+               $scope.explorationTitleService =
+               explorationTitleService;
+               $scope.explorationObjectiveService =
+               explorationObjectiveService;
+               $scope.explorationCategoryService =
+               explorationCategoryService;
+               $scope.explorationLanguageCodeService =
+               explorationLanguageCodeService;
+
+               $scope.requireTitleToBeSpecified = (
+                 !explorationTitleService.savedMemento);
+               $scope.requireObjectiveToBeSpecified = (
+                 !explorationObjectiveService.savedMemento);
+               $scope.requireCategoryToBeSpecified = (
+                 !explorationCategoryService.savedMemento);
+               $scope.requireLanguageToBeSpecified = (
+                 !explorationLanguageCodeService.savedMemento);
+
+               $scope.metadataList = [];
+
+               $scope.CATEGORY_LIST_FOR_SELECT2 = [];
+
+               for (var i = 0; i < CATEGORY_LIST.length; i++) {
+                 $scope.CATEGORY_LIST_FOR_SELECT2.push({
+                   id: CATEGORY_LIST[i],
+                   text: CATEGORY_LIST[i]
+                 });
+               }
+
+               var _states = explorationStatesService.getStates();
+               if (_states) {
+                 var categoryIsInSelect2 = $scope.CATEGORY_LIST_FOR_SELECT2
+                 .some(
+                   function(categoryItem) {
+                     return categoryItem.id ===
+                     explorationCategoryService.savedMemento;
+                   }
+                 );
+
+                 if (!categoryIsInSelect2) {
+                   $scope.CATEGORY_LIST_FOR_SELECT2.push({
+                     id: explorationCategoryService.savedMemento,
+                     text: explorationCategoryService.savedMemento
+                   });
+                 }
+               }
+
+               $scope.save = function() {
+                 // If no objective has been specified yet, require the creator
+                 // to specify it.
+                 if ($scope.requireObjectiveToBeSpecified) {
+                   if (!explorationObjectiveService.displayed) {
```

Comment:
Probably should check all the errors before pushing anything into $scope.metadataList. Otherwise, suppose someone saves with the category unspecified. 'objective' will be pushed in, 'title' will be pushed in, then an error will be raised. They fix the category and then save again. You'll then get duplicate objectives etc. in metadataList.